src/hotspot/share/jfr/recorder/stacktrace/jfrAsyncStackTrace.cpp

```diff
@@ -0,0 +1,148 @@
+/*
+ * Copyright (c) 2024, SAP SE. All rights reserved.
+ * DO NOT ALTER OR REMOVE COPYRIGHT NOTICES OR THIS FILE HEADER.
+ *
+ * This code is free software; you can redistribute it and/or modify it
+ * under the terms of the GNU General Public License version 2 only, as
+ * published by the Free Software Foundation.
+ *
+ * This code is distributed in the hope that it will be useful, but WITHOUT
+ * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
+ * FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License
+ * version 2 for more details (a copy is included in the LICENSE file that
+ * accompanied this code).
+ *
+ * You should have received a copy of the GNU General Public License version
+ * 2 along with this work; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301 USA.
+ *
+ * Please contact Oracle, 500 Oracle Parkway, Redwood Shores, CA 94065 USA
+ * or visit www.oracle.com if you need additional information or have any
+ * questions.
+ */
+
+#include "precompiled.hpp"
+#include "jfr/recorder/stacktrace/jfrAsyncStackTrace.hpp"
+#include "jfr/recorder/stacktrace/jfrStackTrace.hpp"
+#include "jfr/utilities/jfrTypes.hpp"
+#include "jfr/recorder/repository/jfrChunkWriter.hpp"
+#include "jfr/recorder/storage/jfrBuffer.hpp"
+#include "runtime/threadCrashProtection.hpp"
+#include "runtime/vframe.inline.hpp"
+
+JfrAsyncStackFrame::JfrAsyncStackFrame(const Method* method, int bci, u1 type, int lineno, const InstanceKlass* ik) :
+  _klass(ik), _method(method), _line(lineno), _bci(bci), _type(type) {}
+
+
+JfrAsyncStackTrace::JfrAsyncStackTrace(JfrAsyncStackFrame* frames, u4 max_frames) :
+  _frames(frames),
+  _nr_of_frames(0),
+  _max_frames(max_frames),
+  _reached_root(false)
+  {}
+
+
+static const size_t min_valid_free_size_bytes = 16;
+
+bool JfrAsyncStackTrace::record_async(JavaThread* jt, const frame& frame) {
+  NoHandleMark nhm;
+
+  assert(jt != nullptr, "invariant");
+  Thread* current_thread = Thread::current_or_null_safe();
+  if (current_thread == nullptr) {
+    return false;
+  }
+  assert(current_thread->in_asgct(), "invariant");
+  assert(jt != current_thread || current_thread->in_asgct(), "invariant");
```

Comment:
The second clause of the second assert is redundant given the first assert.

Reply:
The second assert is redundant.

Reply:
Good point :)